### PR TITLE
Implement dynamic connectivity

### PR DESCRIPTION
**Pull request by valueote**
impl https://github.com/ex-actor/ex-actor/issues/124

@lixin-wei  I’m trying to reuse the previous code, so I just added contact_node to node_list. This way, we don’t need to introduce  new constructors for ActorRegistry or MessageBroker.

There are a couple of problems in the current implementation, such as the temporary mutex around node_list and the serialization/deserialization logic in the I/O threads. I’ll optimize these in the future.
 
What do you think of the overall design? If this is still the right direction, I’ll keep working on it.

### CODE REVIEW

**Comment by lixin-wei** [comment]
As code becomes more and more complicated, locks will become a bug prone. we should try our best to avoid locks.

If you find the locking is not trivial, you can consider making MessageBroker an actor, just like ActorRegistryRequestProcessor, which I also want to do in the future.

**Comment by lixin-wei** [comment]
> except the old constructor sets enable_dynamic_connectivity_ to false and only sends heartbeat messages.

Can you remove the old heartbeat entirely? Even when using old API, we should still use the new gossip under the hood.

To keep the old API working, as I said in the issue, just use the smallest node_id as the first node, other nodes connect to the first node. The old API can be implemented using the new API.


> reflectcpp doesn’t support serializing std::chrono::time_point.

You can save uint64_t directly, no need to use time_point in `last_seen`

**Comment by lixin-wei** [comment]
I see, when receiving message, no need to check whether it's in peer_nodes or not, just execute it.

When sending back the reply, if you don't know peer's address, just pend the message in another place until gossip arrives.

File: src/ex_actor/internal/network.cc [hunk 1]
```diff
@@ -229,28 +476,59 @@ void MessageBroker::HandleReceivedMessage(zmq::multipart_t multi) {
...
+
+exec::task<bool> MessageBroker::WaitNodeAlive(uint32_t node_id, std::chrono::milliseconds timeout) {
+  auto waiter = peer_nodes_.TryRegisterWaiter(node_id, timeout);
+  if (waiter == nullptr) {
+    co_return true;
   }
+
+  co_await waiter->sem.OnDrained();
+  co_return peer_nodes_.Connected(node_id);
 }
```
Comment (lixin-wei) [inline src/ex_actor/internal/network.cc]: Returning a shared_ptr of waiter seems a little bit arcane.

How about write a `PeerNodes::WaitNodeAlive()` and do everything in it? You can expose PeerNode to outside directly, like `message_broker.GetPeerNodes().WaitNodeAlive()` then no need of a proxy call.

Besides NodeInfoManager seems a better name? what do you think?

File: include/ex_actor/internal/network.h [hunk 2]
```diff
@@ -14,33 +14,56 @@
...
 namespace ex_actor {
 struct NodeInfo {
   uint32_t node_id = 0;
   std::string address;
+  friend bool operator==(const NodeInfo& lhs, const NodeInfo& rhs) { return lhs.node_id == rhs.node_id; }
 };
 }  // namespace ex_actor
 
+namespace std {
+// The current implementation can not reuse the same node_id for different address
```
Comment (lixin-wei) [inline include/ex_actor/internal/network.h]: what will happen now, when adding a node with the same node_id with others?

File: include/ex_actor/internal/actor_registry.h [hunk 3]
```diff
@@ -60,7 +65,11 @@ class ActorRegistryRequestProcessor {
                     "Class can't be created by given args and create function");
     }
     if (is_distributed_mode_) {
-      EXA_THROW_CHECK(node_id_to_address_.contains(config.node_id)) << "Invalid node id: " << config.node_id;
+      if (enable_dynamic_connectivity_) {
```
Comment (lixin-wei) [inline include/ex_actor/internal/actor_registry.h]: we should remove checks like `if (enable_dynamic_connectivity_)`. there shouldn't be any difference. New and old API should share exactly the same implementation under the hood.

The old API just delegates to the new API, and calls WaitNodeAlive automatically for users.

File: src/ex_actor/internal/network.cc (hunk 1 above)
Comment (valueote) [reply src/ex_actor/internal/network.cc]: > How about write a PeerNodes::WaitNodeAlive() and do everything in it?

That sounds like a good idea. I’ll try it. And I agree that `NodeInfoManager` is a better name.
I’m kind of bad at naming methods and classes 😂.

File: include/ex_actor/internal/network.h (hunk 2 above)
Comment (valueote) [reply include/ex_actor/internal/network.h]: In the previous `PeerNodes` implementation, I used `NodeInfo` as the map key.Its hash only uses node_id, so we can’t distinguish nodes that share the same node_id but have different addresses.

For now  we can just remove the hash function cause it's useless.

However, note that in the current implementation, If we add a node with the same ID, `node_id_to_send_socket_` won’t update to the new address.

File: include/ex_actor/internal/actor_registry.h (hunk 3 above)
Comment (valueote) [reply include/ex_actor/internal/actor_registry.h]: OK.  I will modify this.

File: include/ex_actor/internal/network.h (hunk 2 above)
Comment (lixin-wei) [reply include/ex_actor/internal/network.h]: I think it's better to throw an exception and crash loudly in this case.

Also could you pls add a test for this corner case?

Comment (valueote) [reply include/ex_actor/internal/network.h]: Sure, I will.

File: include/ex_actor/internal/actor_registry.h [hunk 4]
```diff
@@ -88,6 +89,11 @@ class ActorRegistryBackend {
                 << config.node_id << ", this_node_id=" << this_node_id_ << ", actor_type=" << typeid(UserClass).name();
     }
 
+    if (is_distributed_mode_) {
+      bool connected = co_await message_broker_->WaitNodeAlive(config.node_id, config.remote_creation_timeout);
```
Comment (lixin-wei) [inline include/ex_actor/internal/actor_registry.h]: Ah I meant call WaitNodeAlive in `Init(node_list)`, not here. No need to add `remote_creation_timeout` currently.

when creating actor or sending actor messages, if you find the node is not alive, throw exception immediately.

```cpp
void Init(..., node_list) { // the old API
  Init(..., /*contact_node=*/node_list[0]); // delegate to new API.
  for (node : node_list) {
      bool ok = sync_wait(WaitNodeAlive(node));
      if (!ok) {
           throw ...
      }
  }
}
```

**Comment by valueote** [comment]
@lixin-wei Do you have any thoughts on the current implementation? If everything is fine, I think this PR is ready to be merged.

File: src/ex_actor/internal/network.cc [hunk 5]
```diff
@@ -16,33 +16,253 @@
...
+
+void NodeInfoManager::Add(uint32_t node_id, const NodeState& state) {
+  std::lock_guard lock(mutex_);
+  auto [iter, inserted] = node_id_to_state_.try_emplace(node_id, state);
+  if (!inserted) {
+    iter->second.node_id = state.node_id;
+    iter->second.address = state.address;
+    iter->second.liveness = NodeState::Liveness::kAlive;
+    iter->second.last_seen = GetTimeMs();
+  }
```
Comment (lixin-wei) [inline src/ex_actor/internal/network.cc:60]: why not just overwrite it - `node_id_to_state_[node_id]=state`? seems no difference.

And in which case will there be duplicate node_id? Seems it's already been checked in EstablishConnectionTo

File: src/ex_actor/internal/network.cc [hunk 6]
```diff
@@ -16,33 +16,253 @@
...
+  alive_peers_ += 1;
```
Comment (lixin-wei) [inline src/ex_actor/internal/network.cc]: when node already exists, alive_peers_ should not be increased.

File: src/ex_actor/internal/actor_registry.cc [hunk 7]
```diff
@@ -202,31 +224,88 @@ ActorRegistry::~ActorRegistry() {
...
+    if (node.node_id != this_node_id) {
+      auto [connected] = stdexec::sync_wait(WaitNodeAlive(node.node_id, 4000)).value();
+      EXA_THROW_CHECK(connected) << "Can not connect to the node " << node.node_id;
+    }
+  }
+}
+
+ActorRegistry::ActorRegistry(uint32_t thread_pool_size, std::unique_ptr<TypeErasedActorScheduler> scheduler,
+                             const ClusterConfig& cluster_config)
+    : is_distributed_mode_(),
```
Comment (lixin-wei) [inline src/ex_actor/internal/actor_registry.cc]: seems you forget to set `is_distributed_mode_`? I'm curious why test can pass.

File: src/ex_actor/internal/network.cc [hunk 8]
```diff
@@ -16,33 +16,253 @@
...
+  for (const auto& pair : node_id_to_state_) {
+    log::Info("[PeerNodes] Node {}, address {}, state {} ", pair.first, pair.second.address,
+              static_cast<uint8_t>(pair.second.liveness));
+  }
+}
+
+GossipMessage NodeInfoManager::GenerateGossipMessage() {
+  GossipMessage message;
+  message.node_states.reserve(node_id_to_state_.size());
+  std::lock_guard lock(mutex_);
```
Comment (lixin-wei) [inline src/ex_actor/internal/network.cc:127]: should this lock also cover `node_id_to_state_.size()`? because other threads might be modifying it.

so as the lock in GetHealthyNodeList

File: src/ex_actor/internal/actor_registry.cc [hunk 9]
```diff
@@ -202,31 +224,88 @@ ActorRegistry::~ActorRegistry() {
 
 ActorRegistry::ActorRegistry(uint32_t thread_pool_size, std::unique_ptr<TypeErasedActorScheduler> scheduler,
                              uint32_t this_node_id, const std::vector<NodeInfo>& cluster_node_info,
-                             HeartbeatConfig heartbeat_config)
+                             NetworkConfig network_config)
```
Comment (lixin-wei) [inline src/ex_actor/internal/actor_registry.cc]: Can you remove this old version of constructor from ActorRegistry&ActorRegistryBackend?

Then the glue code can be moved to top-level, i.e. `ex_actor::Init()`. write a util function to avoid duplication.

By this way we can make core core interfaces clean.

File: include/ex_actor/internal/network.h [hunk 10]
```diff
@@ -112,29 +169,59 @@ class MessageBroker {
...
+  void EstablishConnectionTo(const NodeInfo& node_info);
+  void EstablishConnection(const std::vector<NodeInfo>& node_list);
   void PushOperation(TypeErasedSendOperation* operation);
   void SendProcessLoop(const std::stop_token& stop_token);
   void ReceiveProcessLoop(const std::stop_token& stop_token);
   void HandleReceivedMessage(zmq::multipart_t multi);
-  void CheckHeartbeat();
-  void SendHeartbeat();
+  void SendGossip();
+  void SendFirstGossipMessage(const NodeInfo& contact_node);
```
Comment (lixin-wei) [inline include/ex_actor/internal/network.h]: seems no definition, remove it.

File: include/ex_actor/internal/network.h [hunk 11]
```diff
@@ -112,29 +169,59 @@ class MessageBroker {
...
+    std::mutex mutex_;
+  };
+
+  NodeInfo this_node_ {};
   std::function<void(uint64_t received_request_id, ByteBufferType data)> request_handler_;
-  HeartbeatConfig heartbeat_;
+  NetworkConfig network_config_;
   std::atomic_uint64_t send_request_id_counter_ = 0;
   std::atomic_uint64_t received_request_id_counter_ = 0;
+  size_t contact_node_index_ = 0;
```
Comment (lixin-wei) [inline include/ex_actor/internal/network.h]: contact_node_index_ seems not used anywhere?

File: include/ex_actor/internal/network.h [hunk 12]
```diff
@@ -112,29 +169,59 @@ class MessageBroker {
...
   void ReplyRequest(uint64_t received_request_id, ByteBufferType data);
 
+  bool CheckNodeConnected(uint32_t node_id);
+
+  exec::task<bool> WaitNodeAlive(uint32_t node_id, uint64_t timeout_ms);
+
  private:
-  void EstablishConnections();
+  void EstablishConnectionTo(const NodeInfo& node_info);
+  void EstablishConnection(const std::vector<NodeInfo>& node_list);
```
Comment (lixin-wei) [inline include/ex_actor/internal/network.h]: EstablishConnection seems not used anywhere

File: src/ex_actor/internal/network.cc [hunk 13]
```diff
@@ -16,33 +16,253 @@
...
+    node_id_to_waiters_.erase(it);
+  }
+
+  for (auto& waiter : waiters) {
+    waiter->arrive.store(true, std::memory_order_release);
+    waiter->sem.Acquire(1);
+  }
+}
+
+void NodeInfoManager::NotifyAllWaiters() {
```
Comment (lixin-wei) [inline src/ex_actor/internal/network.cc]: NotifyAllWaiters seems not called by anyone.

File: src/ex_actor/internal/actor_registry.cc [hunk 14]
```diff
@@ -202,31 +224,88 @@ ActorRegistry::~ActorRegistry() {
...
-            heartbeat_config);
+            });
       }()),
       backend_actor_(scheduler_->Clone(), ActorConfig {.node_id = this_node_id_}, scheduler_->Clone(), this_node_id,
                      cluster_node_info, message_broker_.get()),
+      backend_actor_ref_(this_node_id_, this_node_id_, /*actor_id=*/UINT64_MAX, &backend_actor_,
+                         message_broker_.get()) {
+  for (const auto& node : cluster_node_info) {
+    if (node.node_id != this_node_id) {
+      auto [connected] = stdexec::sync_wait(WaitNodeAlive(node.node_id, 4000)).value();
```
Comment (lixin-wei) [inline src/ex_actor/internal/actor_registry.cc]: move 4000 to constant.h

File: src/ex_actor/internal/network.cc [hunk 15]
```diff
@@ -145,32 +391,39 @@ void MessageBroker::SendProcessLoop(const std::stop_token& stop_token) {
...
+        multi.addmem(serialized_identifier.data(), serialized_identifier.size());
+        multi.add(std::move(operation->data));
+        auto& send_socket = node_id_to_send_socket_.At(operation->identifier.response_node_id);
+        EXA_THROW_CHECK(multi.send(send_socket));
+        if (operation->identifier.flag == MessageFlag::kQuit || operation->identifier.flag == MessageFlag::kGossip) {
+          operation->Complete(ByteBufferType {});
+        }
+        any_item_pulled = true;
+      } else {
+        node_id_to_pending_request_.Add(response_node_id, operation);
```
Comment (lixin-wei) [inline src/ex_actor/internal/network.cc]: we should not pend outgoing request sliently when node is not alive. Instead propogate an exception to caller side(by setting errors to this sender).

the incoming request can be processed and reply can be pended, that's fine. Because the node must exists.

DeferredOperations can be a non-template after removing this.

please add a test case for this exception throwing.

File: src/ex_actor/internal/network.cc [hunk 16]
```diff
@@ -145,32 +391,39 @@ void MessageBroker::SendProcessLoop(const std::stop_token& stop_token) {
...
+      if (node_id_to_send_socket_.Contains(request_node_id)) {
+        auto& send_socket = node_id_to_send_socket_.At(request_node_id);
+        auto serialized_identifier = Serialize(reply_operation.identifier);
+        zmq::multipart_t multi;
+        multi.addmem(serialized_identifier.data(), serialized_identifier.size());
+        multi.add(std::move(reply_operation.data));
+        EXA_THROW_CHECK(multi.send(send_socket));
+        any_item_pulled = true;
+      } else {
+        node_id_to_pending_replies_.Add(request_node_id, std::move(reply_operation));
```
Comment (lixin-wei) [inline src/ex_actor/internal/network.cc:377]: a corner case:

1. Send thread: node_id_to_send_socket_.Contains(X) returns false
2. Recv thread: EstablishConnectionTo(X) — inserts socket, drains pending_request_ for X (currently empty)
3. Send thread: node_id_to_pending_request_.Add(X, operation) — queues the operation
4. this request will never be processed.

This is tricky, one possible solution is, check all pending requests periodically in recv loop:

```cpp
void MessageBroker::ReceiveProcessLoop(const std::stop_token& stop_token) {
  while (!stop_token.stop_requested()) {
    ...
    CheckAndDrainPendingRequests().
    ...
  }
}
```

BTW the locks are getting more and more complex now, for next step we can make MessageBroker an actor. see https://github.com/ex-actor/ex-actor/issues/159

File: src/ex_actor/internal/network.cc [hunk 17]
```diff
@@ -145,32 +391,39 @@ void MessageBroker::SendProcessLoop(const std::stop_token& stop_token) {
...
+      }
+    }
     if (!any_item_pulled) {
+      SendGossip();
```
Comment (lixin-wei) [inline src/ex_actor/internal/network.cc]: why only send gossip when no items pulled? this might cause starvation in heavy workflow.

File: src/ex_actor/internal/network.cc [hunk 18]
```diff
@@ -16,33 +16,253 @@
...
+}
+
+bool NodeInfoManager::Contains(uint32_t node_id) {
+  std::lock_guard lock(mutex_);
+  return node_id_to_state_.contains(node_id);
+}
+
+void NodeInfoManager::DeactivateNode(uint32_t node_id) {
+  std::lock_guard lock(mutex_);
+  auto& state = node_id_to_state_.at(node_id);
```
Comment (lixin-wei) [inline src/ex_actor/internal/network.cc:96]: If a quit message arrives from a node that hasn't been registered yet (e.g. gossip hasn't propagated), will this `at` cause crash?

File: src/ex_actor/internal/network.cc (hunk 16 above)
Comment (valueote) [reply src/ex_actor/internal/network.cc:377]: I’m a bit conflicted. I think we should finish implementing the current feature first; switching to another feature midway doesn’t seem like a good idea. But managing too many locks is indeed painful and error-prone. What do you think? Should we finish #159 first?

This PR has spanned a relatively long period, which has led to a lot of redundancy and some obvious mistakes in the code. I apologize for that. In future development, I’ll have an AI agent review my code first to help save us time.

File: src/ex_actor/internal/actor_registry.cc (hunk 7 above)
Comment (valueote) [reply src/ex_actor/internal/actor_registry.cc]: This boolean is false, so the `ActorRegistry` destructor does not call `ClusterAlignedStop()`. However, the `MessageBroker` destructor still calls `ClusterAlignedStop()`. We can remove this variable and, in the `ActorRegistry` destructor, simply check whether message_broker is nullptr.

File: src/ex_actor/internal/actor_registry.cc (hunk 9 above)
Comment (valueote) [reply src/ex_actor/internal/actor_registry.cc]: If so, this PR will introduce a breaking change because some existing code uses the node_list-based `ActorRegistry` constructor.

File: src/ex_actor/internal/network.cc (hunk 16 above)
Comment (lixin-wei) [reply src/ex_actor/internal/network.cc:377]: Yeah you don't need to make MessageBroker an actor now, I just mention this casually, by saying 'next step' I mean next PR.

File: src/ex_actor/internal/actor_registry.cc (hunk 9 above)
Comment (lixin-wei) [reply src/ex_actor/internal/actor_registry.cc]: You are right, let's break ActorRegistry. Distributed mode is still experimental and we didn't promise API stability, let along we recommend using Init() instead of ActorRegistry now. It's a trade-off, and I think it worth it to break a rarely used API to make code clean. Not breaking Init() is more than enough.

File: src/ex_actor/internal/network.cc (hunk 16 above)
Comment (lixin-wei) [reply src/ex_actor/internal/network.cc:377]: > I’ll have an AI agent review my code first to help save us time.

That's great, I recommend cursor+opus 4.6, which is very powerful.

File: src/ex_actor/internal/network.cc (hunk 18 above)
Comment (valueote) [reply src/ex_actor/internal/network.cc:96]: It's safe here, we will call `RefreshLastSeen`  for every message, and `RefreshLastSeen` will insert unknown nodes into `node_id_to_state_` .

File: src/ex_actor/internal/network.cc [hunk 19]
```diff
@@ -145,31 +325,48 @@ void MessageBroker::SendProcessLoop(const std::stop_token& stop_token) {
...
+        auto& send_socket = node_id_to_send_socket_.At(operation->identifier.response_node_id);
+        EXA_THROW_CHECK(multi.send(send_socket));
+        if (operation->identifier.flag == MessageFlag::kQuit || operation->identifier.flag == MessageFlag::kGossip) {
+          operation->Complete(ByteBufferType {});
+        }
+        any_item_pulled = true;
+      } else {
+        EXA_THROW << fmt_lib::format("Node {} is trying to send request to unconnected node {}", this_node_.node_id,
+                                     response_node_id);
       }
```
Comment (lixin-wei) [inline src/ex_actor/internal/network.cc]: this will crash the program IIUC. You should add a new interface called SetError to SendRequestOperation, call `receiver.set_error()`. could you please add a test to cover this?

File: src/ex_actor/internal/network.cc [hunk 20]
```diff
@@ -16,33 +16,218 @@
...
-                             std::function<void(uint64_t received_request_id, ByteBufferType data)> request_handler,
-                             HeartbeatConfig heartbeat_config)
-    : node_list_(std::move(node_list)),
-      this_node_id_(this_node_id),
+NodeInfoManager::NodeInfoManager(uint32_t this_node_id) : this_node_id_(this_node_id) {}
+
+void NodeInfoManager::Add(uint32_t node_id, const NodeState& state) {
+  std::lock_guard lock(mutex_);
+  node_id_to_state_[node_id] = state;
+  alive_peers_ += 1;
```
Comment (lixin-wei) [inline src/ex_actor/internal/network.cc]: NIT: add a denfensive check to ensure node doesn't exist. (I know there is a check outside, just make every method self-contained safe.

File: src/ex_actor/internal/network.cc (hunk 19 above)
Comment (valueote) [reply src/ex_actor/internal/network.cc]: Should we pass the following exception to SetError?
https://github.com/ex-actor/ex-actor/blob/5ba27cb86b5ac3c11a5d351d9394a705ebcfc7a2/src/ex_actor/internal/network.cc#L154

https://github.com/ex-actor/ex-actor/blob/5ba27cb86b5ac3c11a5d351d9394a705ebcfc7a2/src/ex_actor/internal/network.cc#L169

Comment (lixin-wei) [reply src/ex_actor/internal/network.cc]: No, it's framework-side fault, not user's fault, crash is expected.

File: include/ex_actor/internal/network.h [hunk 21]
```diff
@@ -124,6 +124,9 @@ class MessageBroker {
     virtual void Complete(ByteBufferType /*response_data*/) {
       EXA_THROW << "TypeErasedOperation::Complete should not be called";
     }
+    virtual void SetError(std::exception_ptr /*error*/) {
```
Comment (lixin-wei) [inline include/ex_actor/internal/network.h]: can we make these 2 interfaces pure-virtual? I forget why I wrote a throw here.

File: test/dynamic_connectivity_test.cc [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+    } else {
+      target_node_id = cluster_size_ - 1 - this_node_id_;
+      if (target_node_id == this_node_id_) {
+        target_node_id = (this_node_id_ + 1) % cluster_size_;
+      }
+    }
+
+    auto [connected] = stdexec::sync_wait(ex_actor::WaitNodeAlive(target_node_id, 6000)).value();
+    EXA_THROW_CHECK(connected) << fmt::format("Error: node {} can't connected to node {}", this_node_id_,
+                                              target_node_id);
```
[the remark this answers is not in the document]
Comment (lixin-wei) [reply test/dynamic_connectivity_test.cc]: @valueote this sounds resonable

**Comment by valueote** [comment]
@lixin-wei   I've already added a test case for the send exception. Is that sufficient?

https://github.com/valueote/ex-actor/blob/b53c32eb554e6c80cd6a021e3bc3a476c600fa5b/test/network_test.cc#L223-L235

File: test/network_test.cc [hunk 23]
```diff
@@ -45,3 +154,82 @@ TEST(NetworkTest, MessageBrokerTest) {
...
+    std::jthread node_0(create_node, nodes.at(0));
+    std::jthread node_1(create_node, nodes.at(1), nodes.at(0));
+    std::jthread node_2(create_node, nodes.at(2), nodes.at(0));
+  };
+
+#ifdef _WIN32
+  EXPECT_DEATH(duplicate_node_test(), "");
+#else
+  EXPECT_DEATH(duplicate_node_test(), "Nodes with the same node ID but different addresses exist in the cluster");
+#endif  // _WIN32
```
Comment (lixin-wei) [inline test/network_test.cc:206]: ~~what's these used for?~~. i see

**Review by lixin-wei — APPROVED** [review]

**Comment by lixin-wei** [comment]
@valueote merged, thank you so much for your effort over the past months, I really appreciate it. It's also very impressive that you can finish such a complex feature by yourself. Good job!

For next I wonder whether you are interested in https://github.com/ex-actor/ex-actor/issues/159?
